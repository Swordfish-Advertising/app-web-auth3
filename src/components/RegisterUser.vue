<template>
  <v-form ref="form" v-model="validForm">
    <h1>Register a new user</h1>

    <v-text-field
      label="E-mail"
      :rules="[rules.required, rules.email]"
    ></v-text-field>

    <v-text-field
      label="Username"
      :rules="[rules.required]"
    ></v-text-field>
    
    <Password :confirmation="true"></Password>

    <v-autocomplete
      label="Hosting"
      :items="hosts"
      placeholder="Choose where to store your data..."
      :rules="[rules.required]"
    ></v-autocomplete>
    
    <div>
      By registering you agree with our
      <a target="_blank" href="https://pryv.com/terms-of-use/">terms of use</a>.
    </div>
    
    <v-btn
      @click="submit"
      :disabled="!validForm"
    >Submit</v-btn>

    <v-btn
      @click="clear"
    >Clear</v-btn>

    <NavigationButton
      :title="'Already a Pryv user ? Sign in.'"
      :page="'auth'"
    ></NavigationButton>

  </v-form>
</template>

<script>
  import Password from './Password.vue'
  import NavigationButton from './NavigationButton.vue'

  export default {
    components: {
      Password,
      NavigationButton
    },
    data: () => ({
      hosts: [ 'EU', 'US', 'ASIA' ],
      rules: {
        required: value => !!value || 'This field is required.',
        email: value => /.+@.+/.test(value) || 'E-mail must be valid'
      },
      validForm: false
    }),
    methods: {
      submit () {
        if (this.$refs.form.validate()) {
          alert('Submit form...')
        }
      },
      clear () {
        this.$refs.form.reset()
      }
    }
  }
</script>