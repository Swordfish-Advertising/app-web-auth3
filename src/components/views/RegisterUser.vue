<template>
  <v-form id="registerForm" ref="form" v-model="validForm">

    <v-alert
      :value="err"
      type="error"
      transition="scale-transition"
    >{{err}}</v-alert>

    <h1>Register a new user</h1>

    <v-text-field
      label="E-mail"
      id="email"
      v-model="email"
      :rules="[rules.required, rules.email]"
    ></v-text-field>

    <v-text-field
      label="Username"
      id="username"
      v-model="username"
      :rules="[rules.required]"
    ></v-text-field>
    
    <Password v-model="password" :confirmation="true"></Password>

    <v-autocomplete
      label="Hosting"
      id="hosting"
      v-model="hosting"
      :items="hosts"
      placeholder="Choose hosting..."
      :rules="[rules.required]"
    ></v-autocomplete>

    <div>
      By registering you agree with our
      <a target="_blank" href="https://pryv.com/terms-of-use/">terms of use</a>.
    </div>
    
    <v-btn
      id="submitButton"
      @click="submit"
      :disabled="!validForm"
    >Submit</v-btn>

    <v-btn
      id="clearButton"
      @click="clear"
    >Clear</v-btn>

    <router-link :to="{ name: 'Authorization' }">Already a Pryv user ? Sign in.</router-link>

  </v-form>
</template>

<script>
  import Password from './bits/Password.vue';
  import Pryv from '../models/Pryv.js';
  import context from '../../context.js';

  export default {
    components: {
      Password,
    },
    data: () => ({
      username: '',
      password: '',
      email: '',
      hosting: '',
      hosts: [],
      err: '',
      rules: {
        required: value => !!value || 'This field is required.',
        email: value => /.+@.+/.test(value) || 'E-mail must be valid.'
      },
      validForm: false
    }),
    async created() {
      context.pryv.setErrorHandler(err => {
        return this.err = JSON.stringify(err);
      });
      this.hosts = await context.pryv.getAvailableHostings();
    },
    methods: {
      async submit () {
        if (this.$refs.form.validate()) {
          await context.pryv.createUser(
            this.username,
            this.password,
            this.email,
            this.hosting
          );
          this.$router.push('auth');
        }
      },
      clear () {
        this.$refs.form.reset()
      }
    }
  }
</script>