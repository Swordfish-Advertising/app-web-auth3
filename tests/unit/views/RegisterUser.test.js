import { shallowMount, RouterLinkStub } from '@vue/test-utils';
import RegisterUser from '@/components/views/RegisterUser';
import Vue from 'vue';
import Vuetify from 'vuetify';

Vue.use(Vuetify);

describe('RegisterUser.test.js', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(RegisterUser, {
      mocks: {
        $route: {query: {}},
      },
      stubs: {
        RouterLink: RouterLinkStub,
      },
    });
  });

  it('renders correctly (snapshots matching)', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('deactivates submit button when form is not yet filled', () => {
    const submitButton = wrapper.find('#submitButton');
    expect(wrapper.vm.validForm).toBe(false);
    expect(submitButton.attributes().disabled).toBe('true');
  });

  it('activates submit button when form is valid', async () => {
    wrapper.setData({
      validForm: true,
    });
    await Vue.nextTick();
    const submitButton = wrapper.find('#submitButton');
    expect(submitButton.attributes().disabled).toBeFalsy();
  });

  it('validates email correctly', () => {
    const emailRule = wrapper.vm.rules.email;
    expect(emailRule('invalidEmail')).toBe('E-mail must be valid.');
    expect(emailRule('validEmail@test.com')).toBe(true);
  });

  it('validates required fields correctly', () => {
    const requiredRule = wrapper.vm.rules.required;
    expect(requiredRule()).toBe('This field is required.');
    expect(requiredRule('notEmpty')).toBe(true);
  });
});
