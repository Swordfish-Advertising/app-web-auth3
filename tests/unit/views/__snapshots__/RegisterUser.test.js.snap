// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterUser.test.js renders correctly (snapshots matching) 1`] = `
<vform-stub
  id="registerForm"
>
  <valert-stub
    transition="scale-transition"
    type="error"
  >
    
  </valert-stub>
   
  <h1>
    Register a new user
  </h1>
   
  <vtextfield-stub
    id="email"
    label="E-mail"
    rules="function required(value) {
          return !!value || 'This field is required.';
        },function email(value) {
          return (/.+@.+/.test(value) || 'E-mail must be valid.'
          );
        }"
  />
   
  <vtextfield-stub
    id="username"
    label="Username"
    rules="function required(value) {
          return !!value || 'This field is required.';
        }"
  />
   
  <password-stub
    confirmation="true"
    value=""
  />
   
  <vautocomplete-stub
    allowoverflow="true"
    browserautocomplete="off"
    filter="function _default(item, queryText, itemText) {
                var hasValue = function hasValue(val) {
                    return val != null ? val : '';
                };
                var text = hasValue(itemText);
                var query = hasValue(queryText);
                return text.toString().toLowerCase().indexOf(query.toString().toLowerCase()) > -1;
            }"
    id="hosting"
    items=""
    label="Hosting"
    menuprops="[object Object]"
    placeholder="Choose hosting..."
    rules="function required(value) {
          return !!value || 'This field is required.';
        }"
  />
   
  <div>
    
    By registering you agree with our
    
    <a
      href="https://pryv.com/terms-of-use/"
      target="_blank"
    >
      
    terms of use
    </a>
    .
  
  </div>
   
  <vbtn-stub
    activeclass="v-btn--active"
    disabled="true"
    id="submitButton"
    ripple="true"
    tag="button"
    type="button"
  >
    Submit
  </vbtn-stub>
   
  <vbtn-stub
    activeclass="v-btn--active"
    id="clearButton"
    ripple="true"
    tag="button"
    type="button"
  >
    Clear
  </vbtn-stub>
</vform-stub>
`;
